### ISSUE TITLE
`F821` and `F841` false negatives with capturing exceptions occurred in except clause

### ISSUE THREAD

**Issue by jonathandung** [astral-sh/ruff#25097]
### Summary

Neither type checkers nor linters are good at recognizing variables becoming unbound when used as the target of an except clause capture.

```python
# _0.py
e = None
print(e)
try:
    0/1
except Exception as e: ... # should be F841 here, just like
print(e)
try:
    1/0
except ZeroDivisionError as e: ... # here (I don't know what the difference is)
print(e) # no F821 here
del e
```

```bash
> uvx ruff check --isolated
F841 [*] Local variable `e` is assigned to but never used
  --> _0.py:10:29
   |
 8 | try:
 9 |     1/0
10 | except ZeroDivisionError as e: ... # here
   |                             ^
11 | print(e) # there should be F821 here, just like
12 | del e # here
   |
help: Remove assignment to unused variable `e`

F821 Undefined name `e`
  --> _0.py:12:5
   |
10 | except ZeroDivisionError as e: ... # here
11 | print(e) # there should be F821 here, just like
12 | del e # here
   |     ^
   |

Found 2 errors.
[*] 1 fixable with the `--fix` option.
> python _0.py
None
None
Traceback (most recent call last):
  File "<python-input-0>", line 11, in <module>
    print(e) # no F821 here
          ^
NameError: name 'e' is not defined
```

### Version

ruff 0.15.12 (66f93cf7e 2026-04-24)

**Comment by MichaReiser** [comment]
[The `except` documentation](https://docs.python.org/3/reference/compound_stmts.html#except) covers this:

```py
except E as N:
    foo
```

gets translated to

```py
except E as N:
    try:
        foo
    finally:
        del N
```

Meaning, `N` only gets deleted when the exception is raised. Which is why the second print after the `ZeroDevisionError` raises, but the first print does not. 


ty detects the problematic use of `e` when enabling the `possibly-unresolved-reference` rule [https://play.ty.dev/5a1d403a-86cf-4062-9a1f-7eb6c9e34dd5](Playground). 

I think Ruff's behavior here is fine, because the name is sometimes bound, and flagging all uses of `e` could make `F821` too noisy. The alternative is to introduce a new rule only to detect this, but I feel like we should hint users towards ty's `possibly-unresolved-reference`.

**Comment by jonathandung** [comment]
What I'm saying is that:
1. ruff only notices that `e` is not defined when deleting it, but not when printing it on the immediately preceding line.
2. ruff knows that `e` is an unused local the second time, but not the first time; it doesn't matter if `e` is cleared because the `except` clause is empty and does not access `e`

**Comment by MichaReiser** [comment]
> ruff only notices that e is not defined when deleting it, but not when printing it on the immediately preceding line.

Oh, I see. Ruff reports `F821` on the last `del` statement but not on the preceding `print`. This feels inconsistent ([Playground](https://play.ruff.rs/7bc5b120-4fd0-436f-bf5c-8291c86f3c16))

> ruff knows that e is an unused local the second time, but not the first time; it doesn't matter if e is cleared because the except clause is empty and does not access e


I agree that F841 should be raised consistently. I suspect it's related to the `del` difference above

**Comment by hardiksingh5864** [comment]
I’d like to work on this if nobody else is currently handling it.

**Comment by jonathandung** [comment]
Sure

**Comment by MichaReiser** [comment]
The core work here will be to propose what the consistent behavior should be and why

**Comment by jonathandung** [comment]
Well, because in the first instance of capturing and clearing `e`, it is not used either, so both ought to raise `F841`, and because `e` is indeed undefined, with `del e` being correctly flagged, indicating the rule does have this static capability, `F821` should be signalled for both cases as well.

**Comment by koriyoshi2041** [comment]
I took a quick look at the current handling and I think the least surprising consistent behavior is:

- `F841`: report `except ... as e` when `e` is not used inside that handler body. A later read after the handler should not count as using the exception target, because CPython clears that target only on the path where the handler ran.
- `F821`: keep suppressing ordinary reads after an exception target shadows an outer binding. The name is conditionally unbound, so plain `F821` would be too strong here and overlaps with the possibly-unresolved case you mentioned for ty.
- `del e`: match ordinary reads instead of reporting `F821` only for `del`. That is the inconsistent part in the current output: `print(e)` and `del e` are looking at the same conditionally-cleared name, but only `del` is reported.

So the concrete regression matrix I would expect is: both empty handlers get `F841`; the following `print(e)` and `del e` do not get `F821` from Ruff. If that direction sounds right, I can turn it into a narrow regression/fix.
